### ISSUE TITLE
setContext method implementation and injection scope

### ISSUE THREAD

**Issue by ibrcko** [gremo/nest-winston#122]
Hello,

I have noticed that you haven't implemented "setContext()" method in a similar fashion as it is implemented in [@nestjs/logger-service.](https://github.com/nestjs/nest/blob/master/packages/common/services/logger.service.ts)

Since this is a cool feature, what do you think about adding it to this library?

The implementation itself would look something like this:
```javascript
class WinstonLogger implements LoggerService {
  private context?: string;

  constructor(private readonly logger: Logger) {
  }

  public log(message: any, context?: string) {
    return this.logger.info(message, {context: context || this.context});
  }

  public error(message: any, trace?: string, context?: string): any {
    return this.logger.error(message, {trace, context: context || this.context});
  }

  public warn(message: any, context?: string): any {
    return this.logger.warn(message, {context: context || this.context});
  }

  public debug(message: any, context?: string): any {
    return this.logger.debug(message, {context: context || this.context});
  }

  public verbose(message: any, context?: string): any {
    return this.logger.verbose(message, {context: context || this.context});
  }

  public setContext(context: string) {
    this.context = context;
  }
}
```

The use case for this feature would be to set logging context per incoming request, so we don't have to pass "context" as a parameter every time we are logging something.

Furthermore, it would be nice to have an option to set an injection scope for "WinstonLogger" instance i.e. per request.
Since "WinstonLogger" isn't exported from the module, the solution would be either to actually export it, or to add an option "injectionScope" to "WinstonModuleOptions".

What do you think about this?
Thanks.

**Comment by gremo** [comment]
Seems good to me, the part regarding the context @TrejGun @hamxabaig what do you think?

**Comment by hamxabaig** [comment]
Yeahs sounds good, actually this was the reason i had to choose NestJS logger over `nest-winston` It'd be a good addition

**Comment by gremo** [comment]
Can you elaborate more on the "injection scope", providing some examples maybe?

@ibrcko feel free to open a PR about the context feature.

**Comment by TrejGun** [comment]
LGTM :)

**Comment by ibrcko** [comment]
Great!
I'm pleased that you guys are open to this change. (regarding setContext) 
I will create a PR in the upcoming days.

As far as injection scope is concerned.
If you haven't already, you can take a look at this [article](https://dev.to/santoshyadav198613/a-deep-dive-into-the-nestjs-injection-scope-39ih) where injection scopes are explained quite nicely, and as a bonus, the example used for the demo is LoggerService.

Example use case:
_Goal:_ 
I would like to track every request that enters my app by monitoring logs.

_Execution:_
I would make Logger injection scoped per request and in Controller, I would use Logger.setContext("someRequestId"). 
By doing this, I make sure that every request that enters my app has a separate Logger instance with context of "someRequestId".

_nest-winston implementation:_
First option - injection scope can be configured through options:
`javascript
export function createWinstonProviders(loggerOpts: WinstonModuleOptions): Provider[] {
  return [
    {
      provide: WINSTON_MODULE_PROVIDER,
      useFactory: () => createLogger(loggerOpts),
      scope: loggerOpts.scope,
    },
    {
      provide: WINSTON_MODULE_NEST_PROVIDER,
      useFactory: (logger: Logger) => {
        return new WinstonLogger(logger);
      },
      inject: [WINSTON_MODULE_PROVIDER],
      scope: loggerOpts.scope,
    },
  ];
}

export function createWinstonAsyncProviders(options: WinstonModuleAsyncOptions): Provider[] {
  const providers: Provider[] = [
    {
      provide: WINSTON_MODULE_PROVIDER,
      useFactory: (loggerOpts: LoggerOptions) => createLogger(loggerOpts),
      inject: [WINSTON_MODULE_OPTIONS],
      scope: loggerOpts.scope,
    },
    {
      provide: WINSTON_MODULE_NEST_PROVIDER,
      useFactory: (logger: Logger) => {
        return new WinstonLogger(logger);
      },
      inject: [WINSTON_MODULE_PROVIDER],
      scope: loggerOpts.scope,
    },
  ];
`
Second option - export WinstonLogger and leave injection scope configuration to user:
`javascript
export class WinstonLogger implements LoggerService {
  constructor(private readonly logger: Logger) { }

  public log(message: any, context?: string) {
    return this.logger.info(message, { context });
  }

  public error(message: any, trace?: string, context?: string): any {
    return this.logger.error(message, { trace, context });
  }

  public warn(message: any, context?: string): any {
    return this.logger.warn(message, { context });
  }

  public debug?(message: any, context?: string): any {
    return this.logger.debug(message, { context });
  }

  public verbose?(message: any, context?: string): any {
    return this.logger.verbose(message, { context });
  }
}
`

How does this look to you?

**Comment by TrejGun** [comment]
for sure it is better to set scope only once. lets say i want to use my sessionId from redis. this way i can get sessionId in createWinstonAsyncProviders and use once

another question how this would work with gql?

**Comment by ibrcko** [comment]
I'm not sure about GQL, I haven't used it yet, so I can't give you an answer.

**Comment by TrejGun** [comment]
looks like it would work the same way in nest 7
i will try to check when you are done

**Comment by ibrcko** [comment]
So are you guys okay with the injection scope part of this issue?

If you are okay, which method do you prefer: 
1. Injection scope can be configured through options
2. Export WinstonLogger and leave injection scope configuration to the user

**Comment by TrejGun** [comment]
1

**Comment by MickL** [comment]
I would like to use it without a `@Inject()` decorator: `constructor(private logger: WinstonLogger) { }` and especially have no uppercase. Both would be more like a code style preference and to be conform with Nest.js.

For injection scope i like option 1 way more.

Further I would like to set the context like Nest logger per instance: `this.logger.setContext('AppController')` but what Nest does not provide is a global app-context: Having multiple apps running I would like to set a global context per app so it becomes something like `[AdminMicroService:AppController]`, `[ClientMicroService:AppController]`.

**Comment by MickL** [comment]
@ibrcko Any update on this? :)

**Comment by ibrcko** [comment]
@MickL 
I have been experimenting with NestJS's injection scope and noticed that there are some unexpected effects, so I'm having second thoughts about this feature.

Nevertheless, I will make a PR that will give the developer an option to set injection scope, and then the effects of it are risk he would have to take care of.

**Comment by MickL** [comment]
@ibrcko Could you please point out what you mean by unexpected effects?

**Comment by cesc1802** [comment]
i aslo need know. how to update set context

**Comment by Chup4Chups** [comment]
any update ?

**Comment by gremo** [comment]
I' completely open to suggestions, improvements and PRs... so asking the same? Updates @ibrcko @TrejGun

**Comment by TrejGun** [comment]
@MickL has a good idea about using @Inject

**Comment by MickL** [comment]
What I dont understand is why Nest.JS logs have a context but we cant set the context? @ibrcko Could you please point out what exactly the problem is?

> info - RoutesResolver - FcKoelnAppserverController {}:
> info - RouterExplorer - Mapped {/update, POST} route
> info - RouterExplorer - Mapped {/updateEverything, GET} route
> info - NestApplication - Nest application successfully started
> info - undefined - Connected to Redis

**Comment by ibrcko** [comment]
@gremo @TrejGun @MickL 
Hey guys, I'm sorry about my absence.

Unfortunately, I won't be able to add this improvement to the "nest-winston" logger since I am occupied well over my head.

I'm going to close this issue. 
If you think that it is still relevant even without my engagement, feel free to reopen it.

Once again, sorry for the stall.

**Comment by MickL** [comment]
Cant you leave it open? Maybe someone else picks it up.

**Comment by gremo** [comment]
@MickL reopened

**Comment by javiesses** [comment]
Hi team! Are you planning to add the `setContext()` to your `LoggerService`?

**Comment by fmonper1** [comment]
> Hi team! Are you planning to add the `setContext()` to your `LoggerService`?

I am handling it like this on my controllers, but its a bit messy
```
  constructor(
    private readonly userService: UsersService,
    @Inject(WINSTON_MODULE_NEST_PROVIDER)
    private readonly logger: LoggerService,
  ) {
    (logger as WinstonLogger).setContext(UsersController.name);
  }
```

**Comment by chihabeeddine** [comment]
Is there something to be done with what have been said in this ticket #[7442](https://github.com/nestjs/nest/issues/7442) ?
I'm not sure how to replace  this part without DI [Replacing the Nest logger (also for bootstrapping)](https://github.com/gremo/nest-winston#replacing-the-nest-logger-also-for-bootstrapping)
I was doing this priore to v8

``` js
  constructor(private readonly logger: Logger) {
    logger.setContext([class].name);
  }

  ```
  What's the best way to replace this now ?
